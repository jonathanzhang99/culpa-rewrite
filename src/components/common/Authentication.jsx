import PropTypes from "prop-types";
import React, { useContext, useReducer } from "react";
import { Route, Redirect } from "react-router-dom";

export const AuthContext = React.createContext(null);

function authReducer(state, action) {
  switch (action.type) {
    case "LOGIN_START":
      return {
        ...state,
        isLoading: true,
      };
    case "LOGIN_ERROR":
      return {
        ...state,
        isError: true,
        isLoading: false,
      };

    case "LOGIN_SUCCESS":
      localStorage.setItem("authToken", JSON.stringify(action.payload));
      return {
        ...state,
        isError: false,
        isLoading: false,
        isLoggedIn: true,
        authToken: action.payload,
      };

    case "LOGIN_FAILURE":
      return {
        ...state,
        isError: false,
        isLoading: false,
        isLoggedIn: false,
      };
    case "LOGOUT_SUCCESS":
      localStorage.removeItem("authToken");
      return {
        ...state,
        isError: false,
        isLoading: false,
        isLoggedIn: false,
        authToken: undefined,
      };
    default:
      throw new Error();
  }
}

const propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]).isRequired,
};

/**
 * AuthProvider createa React Context with values such that children elements
 * in the tree can access user auth token and manipulate them
 * (e.g. login, logout)
 *
 */
export function AuthProvider({ children }) {
  const existingToken = JSON.parse(localStorage.getItem("authToken"));

  const [state, dispatch] = useReducer(authReducer, {
    isError: false,
    isLoading: false,
    isLoggedIn: !!existingToken,
    authToken: existingToken,
  });

  // `login` either returns an error or null if a token was succesfully
  // retrieved. The error will always be a string message
  const login = async ({ username, password }) => {
    dispatch({ type: "LOGIN_START" });
    const response = await fetch("/api/auth/login", {
      method: "POST",
      body: JSON.stringify({
        username,
        password,
      }),
      headers: {
        "Content-Type": "application/json",
      },
    });
    try {
      const result = await response.json();

      // Require that the backend sends either a token or an error
      result.error
        ? dispatch({ type: "LOGIN_FAILURE" })
        : dispatch({ type: "LOGIN_SUCCESS", payload: result.token });
      return result;
    } catch (err) {
      dispatch({ type: "LOGIN_ERROR" });
      return err;
    }
  };

  // `logout` simply removes the token from the client. It does NOT
  // expire the token.
  const logout = async () => {
    if (state.authToken) {
      dispatch({ type: "LOGOUT_SUCCESS" });
    }
  };

  return (
    <AuthContext.Provider
      value={{
        ...state,
        login,
        logout,
      }}
    >
      {children}
    </AuthContext.Provider>
  );
}

AuthProvider.propTypes = propTypes;

/**
 * ProtectedRoute allows access to the url only if a user is logged in.
 *
 */
export function ProtectedRoute(props) {
  const { isLoggedIn } = useContext(AuthContext);
  // eslint-disable-next-line react/jsx-props-no-spreading
  return isLoggedIn ? <Route {...props} /> : <Redirect to="/" />;
}
