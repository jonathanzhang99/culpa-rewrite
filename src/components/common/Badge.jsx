import PropTypes from "prop-types";
import React from "react";
import { Image } from "semantic-ui-react";

import bronzeNugget from "icons/bronze-nugget.svg";
import goldNugget from "icons/gold-nugget.svg";
import silverNugget from "icons/silver-nugget.svg";

const propTypesBadge = {
  badgeId: PropTypes.number.isRequired,
};

export default function Badge({ badgeId }) {
  switch (badgeId) {
    case 1: {
      return <Image avatar spaced="left" src={goldNugget} />;
    }
    case 2: {
      return <Image avatar spaced="left" src={silverNugget} />;
    }
    case 3: {
      return <Image avatar spaced="left" src={bronzeNugget} />;
    }
    default: {
      return null;
    }
  }
}

Badge.propTypes = propTypesBadge;
