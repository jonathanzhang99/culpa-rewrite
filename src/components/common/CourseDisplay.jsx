import PropTypes from "prop-types";
import React from "react";
import { Link } from "react-router-dom";
import { Header } from "semantic-ui-react";

const propTypesCourseDisplayName = {
  as: PropTypes.string,
  name: PropTypes.string.isRequired,
  code: PropTypes.string,
};

const propTypesCourseDisplayLink = {
  as: PropTypes.string,
  courseId: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  code: PropTypes.string,
};

const defaultProps = {
  as: "p",
  code: "",
};

export function CourseDisplayName({ as, code, name }) {
  return (
    <Header as={as}>
      {code} {name}
    </Header>
  );
}

export function CourseDisplayLink({ as, courseId, code, name }) {
  return (
    <Link
      to={{
        pathname: `/courses/${courseId}`,
      }}
    >
      <CourseDisplayName as={as} code={code} name={name} />
    </Link>
  );
}

CourseDisplayName.propTypes = propTypesCourseDisplayName;
CourseDisplayName.defaultProps = defaultProps;

CourseDisplayLink.propTypes = propTypesCourseDisplayLink;
CourseDisplayLink.defaultProps = defaultProps;
