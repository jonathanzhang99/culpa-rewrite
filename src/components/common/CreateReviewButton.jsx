/* eslint-disable react/jsx-props-no-spreading */
import PropTypes from "prop-types";
import React from "react";
import { Link } from "react-router-dom";
import { Button } from "semantic-ui-react";

const propTypes = {
  courseId: PropTypes.number,
  professorId: PropTypes.number,
};

const defaultProps = {
  courseId: undefined,
  professorId: undefined,
};

export default function CreateReviewButton({ courseId, professorId, ...rest }) {
  return (
    <Link
      to={{
        pathname: "/review/submit",
        state: { courseId, professorId },
      }}
    >
      <Button {...rest} />
    </Link>
  );
}

CreateReviewButton.propTypes = propTypes;
CreateReviewButton.defaultProps = defaultProps;
