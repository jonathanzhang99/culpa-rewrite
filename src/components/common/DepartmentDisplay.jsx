import PropTypes from "prop-types";
import React from "react";
import { Link } from "react-router-dom";
import { Header } from "semantic-ui-react";

const propTypesDepartmentDisplayName = {
  as: PropTypes.oneOf(["header", "text"]),
  departmentName: PropTypes.string.isRequired,
};

const propTypesDepartmentDisplayLink = {
  as: PropTypes.oneOf(["header", "text"]),
  departmentId: PropTypes.number.isRequired,
  departmentName: PropTypes.string.isRequired,
};

const defaultProps = {
  as: "text",
};

export function DepartmentDisplayName({ as, departmentName }) {
  const DepartmentName = as === "header" ? Header : "span";
  return <DepartmentName> {departmentName} </DepartmentName>;
}

export function DepartmentDisplayLink({ as, departmentId, departmentName }) {
  return (
    <Link to={`/department/${departmentId}`}>
      <DepartmentDisplayName as={as} departmentName={departmentName} />
    </Link>
  );
}

DepartmentDisplayName.propTypes = propTypesDepartmentDisplayName;
DepartmentDisplayName.defaultProps = defaultProps;

DepartmentDisplayLink.propTypes = propTypesDepartmentDisplayLink;
DepartmentDisplayLink.defaultProps = defaultProps;
