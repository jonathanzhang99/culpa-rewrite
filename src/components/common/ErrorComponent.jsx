import React from "react";

export default function ErrorComponent() {
  return <div>error occurred</div>;
}
