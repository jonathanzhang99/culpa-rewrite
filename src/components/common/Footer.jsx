import React from "react";
import { Link } from "react-router-dom";
import { Grid, Icon, Image } from "semantic-ui-react";

import culpaIcon from "icons/culpa.svg";

export default function Footer() {
  return (
    <Grid className="footer" columns={2} verticalAlign="middle">
      <Grid.Column textAlign="center">
        <Image centered src={culpaIcon} />
        <Icon color="blue" name="copyright outline" size="small" />
        <span style={{ color: "#82B7E8" }}>1997-2020</span>
      </Grid.Column>
      <Grid.Column textAlign="left">
        <Grid.Row>
          <Link className="footer-link" to="/">
            About
          </Link>
        </Grid.Row>
        <Grid.Row>
          <Link className="footer-link" to="/join-the-team">
            Join The Team
          </Link>
        </Grid.Row>
        <Grid.Row>
          <Icon color="blue" name="mail" size="small" />
          <span className="footer-link">admin@culpa.info</span>
        </Grid.Row>
      </Grid.Column>
    </Grid>
  );
}
