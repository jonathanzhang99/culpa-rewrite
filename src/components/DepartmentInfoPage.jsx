import PropTypes from "prop-types";
import React from "react";
import { useParams } from "react-router-dom";
import { List } from "semantic-ui-react";

import { CourseDisplayLink } from "components/common/CourseDisplay";
import { DepartmentDisplayName } from "components/common/DepartmentDisplay";
import ErrorComponent from "components/common/ErrorComponent";
import LoadingComponent from "components/common/LoadingComponent";
import { ProfessorDisplayLink } from "components/common/ProfessorDisplay";
import useDataFetch from "components/common/useDataFetch";

const propTypes = {
  departmentName: PropTypes.string.isRequired,
  departmentCourses: PropTypes.arrayOf(
    PropTypes.shape({
      courseId: PropTypes.number.isRequired,
      courseName: PropTypes.string.isRequired,
    }).isRequired
  ).isRequired,
  departmentProfessors: PropTypes.arrayOf(
    PropTypes.shape({
      professorId: PropTypes.number.isRequired,
      firstName: PropTypes.string.isRequired,
      lastName: PropTypes.string.isRequired,
    }).isRequired
  ).isRequired,
};

export function DepartmentInfo({
  departmentCourses,
  departmentName,
  departmentProfessors,
}) {
  return (
    // TODO: Add styling into two columns with alphabetical sections
    <div>
      <DepartmentDisplayName as="header" departmentName={departmentName} />
      <List>
        {departmentCourses.map(({ courseId, courseName }) => {
          return (
            <List.Item key={courseId}>
              <CourseDisplayLink courseId={courseId} courseName={courseName} />
            </List.Item>
          );
        })}
      </List>
      <List>
        {departmentProfessors.map(({ professorId, firstName, lastName }) => {
          return (
            <List.Item key={professorId}>
              <ProfessorDisplayLink
                firstName={firstName}
                lastName={lastName}
                professorId={professorId}
              />
            </List.Item>
          );
        })}
      </List>
    </div>
  );
}

export default function DepartmentInfoPage() {
  const { departmentId } = useParams();
  const {
    data: { departmentName, departmentCourses, departmentProfessors },
    isLoading,
    isError,
  } = useDataFetch(`/api/department/${departmentId}`, {
    departmentName: "",
    departmentCourses: [],
    departmentProfessors: [],
  });

  if (isLoading || isError) {
    return isLoading ? <LoadingComponent /> : <ErrorComponent />;
  }

  return (
    <DepartmentInfo
      departmentCourses={departmentCourses}
      departmentName={departmentName}
      departmentProfessors={departmentProfessors}
    />
  );
}

DepartmentInfo.propTypes = propTypes;
