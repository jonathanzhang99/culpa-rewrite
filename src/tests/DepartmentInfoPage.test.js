import { render } from "@testing-library/react";
import React from "react";
import { MemoryRouter } from "react-router-dom";

import { AuthContext } from "components/common/Authentication";
import DepartmentInfoPage from "components/DepartmentInfoPage";

describe("DepartmentInfoPage Component", () => {
  const loginSuccess = jest.fn(() => {
    Promise.resolve();
  });
  const testCases = [
    {
      testName: "renders no info",
      departmentName: "",
      departmentCourses: [{}],
      departmentProfessors: [{}],
    },
    {
      testName: "renders department info",
      departmentName: "Computer Science",
      departmentCourses: [
        {
          courseId: 1,
          courseName: "User Interface Design",
        },
        {
          courseId: 2,
          courseName: "Machine Learning",
        },
      ],
      departmentProfessors: [
        {
          professorId: 1,
          firstName: "Lydia",
          lastName: "Chilton",
        },
        {
          professorId: 2,
          firstName: "Nakul",
          lastName: "Verma",
        },
      ],
    },
  ];
  testCases.forEach(
    ({ testName, departmentName, departmentCourses, departmentProfessors }) => {
      test(testName, () => {
        const snapshot = render(
          <MemoryRouter>
            <AuthContext.Provider value={{ login: loginSuccess }}>
              <DepartmentInfoPage
                departmentCourses={departmentCourses}
                departmentName={departmentName}
                departmentProfessors={departmentProfessors}
              />
            </AuthContext.Provider>
          </MemoryRouter>
        );
        expect(snapshot).toMatchSnapshot();
      });
    }
  );
});
