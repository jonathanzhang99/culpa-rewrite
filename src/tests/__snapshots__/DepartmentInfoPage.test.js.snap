// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DepartmentInfo Component renders department info 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui header"
      >
         
        Computer Science
         
      </div>
      <div
        class="ui two column grid"
      >
        <div
          class="column"
        >
          <div
            class="row"
          >
            <div
              class="ui header"
            >
              Courses
            </div>
          </div>
          <div
            class="row"
          >
            <a
              href="/course/1"
            >
              <span>
                
                 
                User Interface Design
              </span>
            </a>
          </div>
          <div
            class="row"
          >
            <a
              href="/course/2"
            >
              <span>
                
                 
                Machine Learning
              </span>
            </a>
          </div>
        </div>
        <div
          class="column"
        >
          <div
            class="row"
          >
            <div
              class="ui header"
            >
              Professors
            </div>
          </div>
          <div
            class="row"
          >
            <a
              href="/professor/1"
            >
              <span>
                Lydia
                 
                Chilton
              </span>
            </a>
          </div>
          <div
            class="row"
          >
            <a
              href="/professor/2"
            >
              <span>
                Nakul
                 
                Verma
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui header"
    >
       
      Computer Science
       
    </div>
    <div
      class="ui two column grid"
    >
      <div
        class="column"
      >
        <div
          class="row"
        >
          <div
            class="ui header"
          >
            Courses
          </div>
        </div>
        <div
          class="row"
        >
          <a
            href="/course/1"
          >
            <span>
              
               
              User Interface Design
            </span>
          </a>
        </div>
        <div
          class="row"
        >
          <a
            href="/course/2"
          >
            <span>
              
               
              Machine Learning
            </span>
          </a>
        </div>
      </div>
      <div
        class="column"
      >
        <div
          class="row"
        >
          <div
            class="ui header"
          >
            Professors
          </div>
        </div>
        <div
          class="row"
        >
          <a
            href="/professor/1"
          >
            <span>
              Lydia
               
              Chilton
            </span>
          </a>
        </div>
        <div
          class="row"
        >
          <a
            href="/professor/2"
          >
            <span>
              Nakul
               
              Verma
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`DepartmentInfo Component renders department with no courses or professors 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui header"
      >
         
        Computer Science
         
      </div>
      <div
        class="ui two column grid"
      >
        <div
          class="column"
        >
          <div
            class="row"
          >
            <div
              class="ui header"
            >
              Courses
            </div>
          </div>
        </div>
        <div
          class="column"
        >
          <div
            class="row"
          >
            <div
              class="ui header"
            >
              Professors
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui header"
    >
       
      Computer Science
       
    </div>
    <div
      class="ui two column grid"
    >
      <div
        class="column"
      >
        <div
          class="row"
        >
          <div
            class="ui header"
          >
            Courses
          </div>
        </div>
      </div>
      <div
        class="column"
      >
        <div
          class="row"
        >
          <div
            class="ui header"
          >
            Professors
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
